### PR TITLE
Updatable inputs

### PR DESCRIPTION
**Pull request by tetron**
Not ready to merge!

This is for tracking work on the feature of being able to safely update input files in-place.

### CODE REVIEW

**Comment by tetron** [comment]
Extends the existing functionality of staging "writable" files to the output directory.  If `inplaceUpdate` is true, files will link (via symlink or bind mount) to the real files and permit updating.

Enforces sequencing of updates by incrementing a "generation" number with each update (*).  Files can only be update by one workflow step at a time, and the updated file must be passed explicitly to downstream steps.  If two steps try to update a file of the same generation it will raise an error.

(*) In writing this, I realized that while write-write conflicts are detected, read-write conflicts are not.  So the lock management probably needs a bit more work.

**Comment by tetron** [comment]
@gijzelaerr this is ready for testing

**Comment by gijzelaerr** [comment]
awesome! Amazing! Thanks!

I'm trying it out now, but I run into troubles running your test/example:
```
gijs@sherlock:~/Work/cwltool (updatable-inputs) ✗
λ  .virtualenv/bin/cwltool tests/wf/updateval_inplace.cwl
.virtualenv/bin/cwltool 1.0.20170307024643
Resolved 'tests/wf/updateval_inplace.cwl' to 'file:///home/gijs/Work/cwltool/tests/wf/updateval_inplace.cwl'
Tool definition failed validation:
tests/wf/updateval_inplace.cwl:3:1: checking field `requirements`
tests/wf/updateval_inplace.cwl:8:3:   checking item
tests/wf/updateval_inplace.cwl:8:3:     Field `class` contains undefined reference to `file:///home/gijs/Work/cwltool/tests/wf/InplaceUpdateRequirement`
```

**Comment by tetron** [comment]
To enable the feature is hidden behind the flag --enable-ext

File: cwltool/job.py [hunk 1]
```diff
@@ -89,8 +89,18 @@ def deref_links(outputs):  # type: (Any) -> None
...
-class CommandLineJob(object):
+def relink_initialworkdir(pathmapper, inplace_update=False):
+    # type: (PathMapper, bool) -> None
+    for src, vol in pathmapper.items():
+        if not vol.staged:
+            continue
+        if vol.type in ("File", "Directory") or (inplace_update and
+                                                 vol.type in ("WritableFile", "WritableDirectory")):
+            if os.path.exists(vol.target):
+                os.remove(vol.target)
```
Comment (gijzelaerr) [inline cwltool/job.py]: this is giving me:
```
Traceback (most recent call last):
  File "/home/gijs/Work/cwltool/cwltool/job.py", line 209, in _execute
    relink_initialworkdir(self.generatemapper, inplace_update=self.inplace_update)
  File "/home/gijs/Work/cwltool/cwltool/job.py", line 99, in relink_initialworkdir
    os.remove(vol.target)
OSError: [Errno 21] Is a directory: '/tmp/tmpC7NZ5I/1484694604_1284.hh_vv.ms'
[job casa_flagdata.cwl] completed permanentFail
{}
```

**Comment by gijzelaerr** [comment]
This fails in the case of a Directory input/output
```
Traceback (most recent call last):
  File "/home/gijs/Work/cwltool/cwltool/job.py", line 209, in _execute
    relink_initialworkdir(self.generatemapper, inplace_update=self.inplace_update)
  File "/home/gijs/Work/cwltool/cwltool/job.py", line 99, in relink_initialworkdir
    os.remove(vol.target)
OSError: [Errno 21] Is a directory: '/tmp/tmpC7NZ5I/1484694604_1284.hh_vv.ms'
[job casa_flagdata.cwl] completed permanentFail
{}
```
here is my CWL file
https://github.com/gijzelaerr/vermeerkat-cwl/blob/99d6e9456f18dc3e0a20f65729b27da6c8b02fb9/cwl/casa_flagdata.cwl

**Comment by tetron** [comment]
Oh, I forgot to even test it for WritableDirectory.  Will take a look.

**Comment by tetron** [comment]
@gijzelaerr try it now, updatable directories should be fixed.

**Comment by mr-c** [comment]
@tetron FYI: I merged in master

**Comment by gijzelaerr** [comment]
I think the merge from master broke something? 
```
 .virtualenv/bin/cwltool --enable-ext --outdir results --cachedir cache cwl/workflow.cwl cwl/job.cwl
.virtualenv/bin/cwltool 1.0.20170307024643
Resolved 'cwl/workflow.cwl' to 'file:///home/gijs/Work/vermeerkat-cwl/cwl/workflow.cwl'
Tool definition failed initialization:
cwl/aoflagger.cwl:15:5: Unsupported requirement InplaceUpdateRequirement
```

**Comment by gijzelaerr** [comment]
yes, f812858 seems to eat the `InplaceUpdateRequirement` just fine.

**Comment by gijzelaerr** [comment]
f812858 gives me
```
 .virtualenv/bin/cwltool --enable-ext --outdir results --cachedir cache cwl/workflow.cwl cwl/job.cwl
.virtualenv/bin/cwltool 1.0.20170310233005
Resolved 'cwl/workflow.cwl' to 'file:///home/gijs/Work/vermeerkat-cwl/cwl/workflow.cwl'
[job curl] Using cached output in /home/gijs/Work/vermeerkat-cwl/cache/e264380f650712a0bed41bd4e101b960
invalid field `_generation`, expected one of: 'class', 'location', 'path', 'basename', 'dirname', 'nameroot', 'nameext', 'checksum', 'size', 'secondaryFiles', 'format', 'contents'
Got workflow error
Traceback (most recent call last):
  File "/home/gijs/Work/cwltool/cwltool/main.py", line 238, in single_job_executor
    r.run(**kwargs)
  File "/home/gijs/Work/cwltool/cwltool/draft2tool.py", line 144, in run
    kwargs.get("compute_checksum", True)), "success")
  File "/home/gijs/Work/cwltool/cwltool/draft2tool.py", line 474, in collect_output_ports
    for r in readers.values():
AttributeError: 'NoneType' object has no attribute 'values'
Workflow error, try again with --debug for more information:
'NoneType' object has no attribute 'values'
```

**Comment by mr-c** [comment]
Oops, sorry. Looks like I did the merge from a stale browser window. I'll fix it now

**Comment by mr-c** [comment]
@tetron I've undone the merge from master with a force push and I'll keep my hands off this PR; mea culpa!

**Comment by tetron** [comment]
@gijzelaerr How does it work if you don't use `--cachedir` ?  Caching results and making destructive updates doesn't really mix...

**Comment by gijzelaerr** [comment]
I'll try! It would still be nice if some of the steps could be cached though. For example the first step in our pipeline is to download a 75 GB file from our data storage and then convert that into an other 75 GB file in a different format. It is a major win to have that cached, but the other ones shouldn't be cached. So there is definitely a case for having mixed cache strategies.

**Comment by gijzelaerr** [comment]
```
  .virtualenv/bin/cwltool --enable-ext --outdir results cwl/workflow.cwl cwl/job.cwl
.virtualenv/bin/cwltool 1.0.20170316183151
Resolved 'cwl/workflow.cwl' to 'file:///home/gijs/Work/vermeerkat-cwl/cwl/workflow.cwl'
Tool definition failed initialization:
Tool definition file:///home/gijs/Work/vermeerkat-cwl/cwl/aoflagger.cwl failed validation:
cwl/aoflagger.cwl:4:1:   checking field `requirements`
cwl/aoflagger.cwl:15:5:     checking item
cwl/aoflagger.cwl:15:5:       Field `class` contains undefined reference to `file:///home/gijs/Work/vermeerkat-cwl/cwl/InplaceUpdateRequirement`
```

**Comment by tetron** [comment]
Ah yes that changed slightly, because it is an extension it needs to be prefixed:

```
$namespaces:
  cwltool: http://commonwl.org/cwltool#
requirements:
  cwltool:InplaceUpdateRequirement:
    inplaceUpdate: true
```

Regarding caching, I remembered that it does record the file size and modified time of each input file as part of the cache (although it currently doesn't descend into directories, which is a bug).  So if a workflow step takes as input a file that was destructively modified later on (updating the mtime) then it won't reuse that step from cache because the input file is different.  On the other hand, if an upstream step produces an output file which is modified in place by a downstream step, that contaminates the output of the upstream step, so the upstream step probably shouldn't be reused either.

Maintaining correctness when applying destructive updates is hard.

**Comment by gijzelaerr** [comment]
Is the namespace to make sure the CWL file will be compatible with other CWL runners?

**Comment by mr-c** [comment]
@gijzelaerr Yes, the namespace makes it very clear that this is an extension from `cwltool` that is not currently part of the released spec. This is akin to the vendor prefixing done nowadays for HTML features in development.

**Comment by gijzelaerr** [comment]
Currently our pipline fails on the last step in our pipeline, where every step modifies the same nested folder structure:
```
[job casa_flagdata] Output of job will be cached in /home/gijs/Work/vermeerkat-cwl/cache/4b4057eef76f16af1cd3ac07b48b7f3b
Exception on step 'casa_flagdata'
[step casa_flagdata] Cannot make job: [job casa_flagdata] wants to modify file:///home/gijs/Work/vermeerkat-cwl/cache/ff7224a8e6539e8a9a0e5809fdbfc8ee/result.ms/table.f7_TSM1 from generation 0 but current generation is 1 (last updated by aoflagger)
[workflow workflow.cwl] outdir is /home/gijs/Work/vermeerkat-cwl/cache/HaVcZw
{}
Final process status is permanentFail
make: *** [run] Error 1
```
I forgot to disable the `--cache` flag, but the cache was empty. Will try again without the `--cache` flag.

**Comment by gijzelaerr** [comment]
oops, turns out that actually is an error in my workflow step, 2 steps modifying the same previous result! so actually it works as expected! It catches this mistake properly, and it also seems to work properly with the cache set.

**Comment by gijzelaerr** [comment]
One last issue, my last step in my workflow fails with this error:

```
[step casa_flagdata] completed success
[job casa_flagdata] Removing input staging directory /tmp/tmpdwLX9x
[job casa_flagdata] Removing temporary directory /tmp/tmpZXH2S3
[workflow workflow.cwl] outdir is /home/gijs/Work/vermeerkat-cwl/cache/yLGyQ7
Copying /home/gijs/Work/vermeerkat-cwl/cache/cb2f046420fff36bd51cce2fe29d0fb6/result.ms to results/result.ms
Unhandled error, try again with --debug for more information:
  [Errno 21] Is a directory: '/home/gijs/Work/vermeerkat-cwl/cache/cb2f046420fff36bd51cce2fe29d0fb6/result.ms'
Traceback (most recent call last):
  File "/home/gijs/Work/cwltool/cwltool/main.py", line 768, in main
    **vars(args))
  File "/home/gijs/Work/cwltool/cwltool/main.py", line 260, in single_job_executor
    kwargs["make_fs_access"](""))
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 243, in relocateOutputs
    stageFiles(pm, moveIt)
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 194, in stageFiles
    stageFunc(p.resolved, p.target)
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 238, in moveIt
    shutil.copy(src, dst)
  File "/usr/lib/python2.7/shutil.py", line 119, in copy
    copyfile(src, dst)
  File "/usr/lib/python2.7/shutil.py", line 82, in copyfile
    with open(src, 'rb') as fsrc:
IOError: [Errno 21] Is a directory: '/home/gijs/Work/vermeerkat-cwl/cache/cb2f046420fff36bd51cce2fe29d0fb6/result.ms'
```

**Comment by gijzelaerr** [comment]
just did a rerun without debug, I think i have the same error

```
[step casa_flagdata] completed success
[workflow workflow.cwl] outdir is /home/gijs/Work/vermeerkat-cwl/cache/R88qhS
Unhandled error, try again with --debug for more information:
  [Errno 21] Is a directory: '/home/gijs/Work/vermeerkat-cwl/cache/cb5bf370951c5f2a7b4538e0e0cb7d41/result.ms'
make: *** [run] Error 1
```

**Comment by tetron** [comment]
@gijzelaerr I'm pretty swamped at the moment, would you be able to look into this?  It looks like right at the end in `relocateOutputs` it assumes it is only copying files, it needs to detect if something is a directory and use `shutil.copytree`.

**Comment by gijzelaerr** [comment]
It seems to be a bit more complex than that. When I add support for recursive dir copy there I get:

```
[workflow workflow.cwl] outdir is /home/gijs/Work/vermeerkat-cwl/cache/kx9MZe
Copying /home/gijs/Work/vermeerkat-cwl/cache/a1f3006db20fae39c82ec21e8297dd1c/result.ms to results/result.ms
Unhandled error, try again with --debug for more information:
  [Errno 2] No such file or directory: u'/result.ms/table.f7_TSM1'
Traceback (most recent call last):
  File "/home/gijs/Work/cwltool/cwltool/main.py", line 768, in main
    **vars(args))
  File "/home/gijs/Work/cwltool/cwltool/main.py", line 260, in single_job_executor
    kwargs["make_fs_access"](""))
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 259, in relocateOutputs
    adjustFileObjs(outputObj, _check_adjust)
  File "/home/gijs/Work/cwltool/cwltool/pathmapper.py", line 41, in adjustFileObjs
    adjustFileObjs(rec[d], op)
  File "/home/gijs/Work/cwltool/cwltool/pathmapper.py", line 41, in adjustFileObjs
    adjustFileObjs(rec[d], op)
  File "/home/gijs/Work/cwltool/cwltool/pathmapper.py", line 44, in adjustFileObjs
    adjustFileObjs(d, op)
  File "/home/gijs/Work/cwltool/cwltool/pathmapper.py", line 39, in adjustFileObjs
    op(rec)
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 256, in _check_adjust
    compute_checksums(fs_access, f)
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 798, in compute_checksums
    with fs_access.open(fileobj["location"], "rb") as f:
  File "/home/gijs/Work/cwltool/cwltool/stdfsaccess.py", line 26, in open
    return open(self._abs(fn), mode)
IOError: [Errno 2] No such file or directory: u'/result.ms/table.f7_TSM1'
```

**Comment by tetron** [comment]
@gijzelaerr how is it working now?

**Comment by gijzelaerr** [comment]
same error, full output here.

https://gist.github.com/gijzelaerr/11115e8ece7ef5e68e4a0823a77a24bd

**Comment by mr-c** [comment]
@gijzelaerr Thanks for the update. Can you try again w/o caching?

**Comment by gijzelaerr** [comment]
same.

```
$   .virtualenv/bin/cwltool --debug --enable-ext --outdir results/ --tmpdir-prefix tmp/ --tmp-outdir-prefix tmp-output/ cwl/workflow.cwl cwl/job.cwl

[...]

[job casa_flagdata] Removing input staging directory /home/gijs/Work/vermeerkat-cwl/tmp/eaDlUr
[job casa_flagdata] Removing temporary directory /home/gijs/Work/vermeerkat-cwl/tmp/ROxfh_
[workflow workflow.cwl] outdir is /home/gijs/Work/vermeerkat-cwl/tmp-output/ORBBrQ
Moving /home/gijs/Work/vermeerkat-cwl/tmp-output/NrCwpP/result.ms to results/result.ms
Unhandled error, try again with --debug for more information:
  [Errno 2] No such file or directory: u'/result.ms/table.f7_TSM1'
Traceback (most recent call last):
  File "/home/gijs/Work/cwltool/cwltool/main.py", line 776, in main
    **vars(args))
  File "/home/gijs/Work/cwltool/cwltool/main.py", line 260, in single_job_executor
    kwargs["make_fs_access"](""))
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 263, in relocateOutputs
    adjustFileObjs(outputObj, _check_adjust)
  File "/home/gijs/Work/cwltool/cwltool/pathmapper.py", line 41, in adjustFileObjs
    adjustFileObjs(rec[d], op)
  File "/home/gijs/Work/cwltool/cwltool/pathmapper.py", line 41, in adjustFileObjs
    adjustFileObjs(rec[d], op)
  File "/home/gijs/Work/cwltool/cwltool/pathmapper.py", line 44, in adjustFileObjs
    adjustFileObjs(d, op)
  File "/home/gijs/Work/cwltool/cwltool/pathmapper.py", line 39, in adjustFileObjs
    op(rec)
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 260, in _check_adjust
    compute_checksums(fs_access, f)
  File "/home/gijs/Work/cwltool/cwltool/process.py", line 802, in compute_checksums
    with fs_access.open(fileobj["location"], "rb") as f:
  File "/home/gijs/Work/cwltool/cwltool/stdfsaccess.py", line 26, in open
    return open(self._abs(fn), mode)
IOError: [Errno 2] No such file or directory: u'/result.ms/table.f7_TSM1'
```

**Comment by tetron** [comment]
@gijzelaerr as a workaround, can you try `--no-compute-checksum` ?

Looks like it is trying to resolve a relative path without having a base directory to start from.

**Comment by gijzelaerr** [comment]
No luck, same issue.

```
$  .virtualenv/bin/cwltool --no-compute-checksum --debug --enable-ext --outdir results/ --tmpdir-prefix tmp/ --tmp-outdir-prefix tmp-output/ cwl/workflow.cwl cwl/job.cwl

[...]
IOError: [Errno 2] No such file or directory: u'/result.ms/table.f7_TSM1'
```

**Comment by mr-c** [comment]
@gijzelaerr Bummer. How hard would it be to make a complete minimal test case with data included?

**Comment by gijzelaerr** [comment]
here you go. https://github.com/gijzelaerr/cwlfail

actually i think I isolated the error a bit more, it only occurs if you specify the `--outdir`

**Comment by tetron** [comment]
@gijzelaerr thanks for the minimal test case, it was extremely helpful.

A bit late on this, but moving Directory outputs to their final location should be much improved now (your test case does the right thing).

**Pull request merged**